src/popover/container.tsx

```diff
@@ -47,145 +48,181 @@ export default function PopoverContainer({
-  }, [position, trackRef, renderWithPortal]);
+
+      // Get the position of the popover relative to the offset parent.
+      const popoverOffset = toRelativePosition(boundingOffset, containingBlockRect);
+
+      // Cache the distance between the trigger and the popover (which stays the same as you scroll),
+      // and use that to recalculate the new popover position.
+      const trackRelativeOffset = toRelativePosition(popoverOffset, toRelativePosition(trackRect, containingBlockRect));
+
+      // Bring back the container to its original position to prevent any flashing.
+      popover.style.top = prevTop;
+      popover.style.left = prevLeft;
+
+      // Allow popover body to scroll if can't fit the popover into the container/viewport otherwise.
+      if (scrollable) {
+        body.style.maxHeight = Math.min(boundingOffset.height, window.innerHeight - parseInt(prevTop)) + 'px';
+        body.style.overflowX = 'hidden';
+        body.style.overflowY = 'auto';
+      }
+
+      // Position the popover
+      internalPositionRef.current = newInternalPosition;
+      setPopoverStyle({ top: popoverOffset.top, left: popoverOffset.left });
+
+      positionHandlerRef.current = () => {
+        const newTrackOffset = toRelativePosition(
+          track.getBoundingClientRect(),
+          containingBlock ? containingBlock.getBoundingClientRect() : viewportRect
+        );
+        setPopoverStyle({
+          top: newTrackOffset.top + trackRelativeOffset.top,
+          left: newTrackOffset.left + trackRelativeOffset.left,
+        });
+      };
+    },
+    [trackRef, keepPositionOnResize, position, renderWithPortal]
+  );
 
   // Recalculate position when properties change.
   useLayoutEffect(() => {
     updatePositionHandler();
   }, [updatePositionHandler, trackKey]);
 
   // Recalculate position when content size changes.
-  useResizeObserver(contentRef, () => updatePositionHandler());
+  useResizeObserver(contentRef, () => {
+    updatePositionHandler(true);
+  });
 
   // Recalculate position on DOM events.
   useLayoutEffect(() => {
     /*
-      This is a heuristic. Some layout changes are caused by user clicks (e.g. toggling the tools panel, submitting a form),
-      and by tracking the click event we can adapt the popover's position to the new layout.
+    This is a heuristic. Some layout changes are caused by user clicks (e.g. toggling the tools panel, submitting a form),
+    and by tracking the click event we can adapt the popover's position to the new layout.
 
-      TODO: extend this to Enter and Spacebar?
+    TODO: extend this to Enter and Spacebar?
     */
-    const updatePosition = () => requestAnimationFrame(() => updatePositionHandler());
+
+    const onClick = (event: UIEvent | KeyboardEvent) => {
+      if (
+        // Do not update position if the click was on the popover itself
+        // and keepPositionOnResize is true.
+        (keepPositionOnResize && nodeContains(popoverRef.current, event.target)) ||
```

Comment:
The click has now nothing to do with resize. Should we introduce a separate flag like keepPositionOnClick?

Reply:
Hm, not very keen to introduce yet another prop, but I get your point. Maybe I'd rather change its name, like `keepPosition`.